### PR TITLE
Aircraft CSV Format Importer

### LINKED ISSUE #935: Aircraft own position CSV import - agree implementation
## 🔗 Feature
#932 

## :floppy_disk: Sample data

We've received the attached CSV file giving a snapshot of position/speed/course information for the "ownship" (file generating) aircraft. 
[OwnPos_UC.csv](https://github.com/debrief/pepys-import/files/6975002/OwnPos_UC.csv)

Viewed sample of file today. It contained 20k rows.  Here is sample with 20 rows:
[Aircraft CSV own track - CSV Output.csv.zip](https://github.com/debrief/pepys-import/files/7014053/Aircraft.CSV.own.track.-.CSV.Output.csv.zip)

## :mag: Exploitation plan

|Date(Uk)       |Time(Z)   |Lat(DD:MM.MM)|Long(DDD:MM.MM)|Lat(DegN)|Long(DegE)       |Altitude(ft)|GndSpeed(kts)|Course(deg)|
|---------------|----------|-------------|---------------|---------|-----------------|------------|-------------|-----------|
|15/03/2020     |10:09:00  |50:37.12N    |005:06.24W     |50.61867 |-5.104           |1215        |81.3         |311        |

|Name           |Sample   |Exploit|Meaning      |Units|Processing       |Target-Table|Target-Field|
|---------------|----------|-------|-------------|-----|-----------------|------------|------------|
|Date(Uk)       |15/03/2020|✓      |Date         |n/a  |Combine date-time|States      |time        |
|Time(Z)        |10:09:00  |✓      |Time         |n/a  |Combine date-time|States      |time        |
|Lat(DD:MM.MM)  |50:37.12N |       |             |     |                 |            |            |
|Long(DDD:MM.MM)|005:06.24W|       |             |     |                 |            |            |
|Lat(DegN)      |50.61867  |✓      |Location-Lat |Degs |Combine Lat/Lng  |States      |location    |
|Long(DegE)     |-5.104    |✓      |Location-Long|Degs |Combine Lat/Lng  |States      |location    |
|Altitude(ft)   |1215      |✓      |Altitude     |ft   |Convert to SI    |States      |elevation   |
|GndSpeed(kts)  |81.3      |✓      |Ground Speed |kts  |Convert to  SI   |States      |speed       |
|Course(deg)    |311       |✓      |Course       |Degs |Convert to Rads  |States      |course      |

Will need to request other data from user, if it's not present in file/folder name.

## Import tests

| Test | Strategy |
| -- | -- |
| **Constructor args** | |
| name | `Merlin CSV` |
| validation_level | `Enhanced` |
| short_name | `Merlin CSV`  |
| datafile_type | `*.csv` |
| **Import tests** | |
| can_load_this_type | `csv` |
| can_load_this_filename |  |
| can_load_this_header |  Contains `Data(Uk)` |
| can_load_this_file |  |

## :confused: Idiosnycracies

For the foreseeable future this data format will come from UK helos.  We can provide these defaults in the import process.

## :information_desk_person: Points of contact / SMEs.

AC via MW.

##  :ok: Acceptance criteria
- [ ] Approach to importing the own position file identified

## ✅  Steps to writing importer
* [x] Familiarise with [Importer development guide](https://pepys-import.readthedocs.io/en/latest/importer_dev_guide.html)
* [x] Familiarise with exploitation strategy (above) and the tables/columns in the [Pepys Schema](https://docs.google.com/drawings/d/e/2PACX-1vRDVCtMH00mwGDLENMOzBq-EcMAzPzJ0vQ9HHmWXa1bVrMFawyqi63Mdf7429DLAUA-_pSHFdiTzR8D/pub?w=2342&h=1575)
* [x] Find an existing importer that looks similar (ideally a CSV-based one)
* [ ] Create new importer & skeleton unit test
* [ ] Obtain an unclass data file, store in repo (`tests\sample_data\track_files`)
* [ ] Develop importer one unit test at a time, against above sample data file
* [ ] Note: if you switch on `import highlighting` in the importer - you can use the highlighted import file to verify things are working as expected

### PR DESCRIPTION
**Pull request by SamWilliams-dev**
## 🧰 Issue
 Fixes #935

## 🚀 Overview: 

This is the implementation of the Aircraft CSV Format Importer which is derived from the base Importer class. This takes in a CSV file with data from a set aircraft entity and imports the data into Pepys if the data is within the correct format. There is significant error handling in the importer to prevent a CSV file with incorrect formatting or incorrect data from being uploaded into Pepys.

## 🤔 Reason: 

There is a requirement from the user to be able to import data from a given aircraft entity using `pepys_import`. Further explanation and understanding of the issue can be found with issue #935

- [x] Tests pass

## Confirmations

- [x] I have chosen reviewers for my PR.
- [x] I have chosen an appropriate label for the PR, adding `interactive_review` if reviewers will need to see UI
- [ ] I have extended/updated the documentation in `\docs` folder
- [ ] Any database content changes (Create, Edit, Delete) are recorded in the Log/Changes tables
- [ ] Any database schema changes are implemented via `alembic revision` [transitions](https://pepys-import.readthedocs.io/en/latest/database_migration.html#how-to-use-it-for-developers)
- [x] I have completed the mandatory sections of this document.
- [x] I have deleted any unused sections.

### CODE REVIEW

File: tests/test_load_aircraft_csv.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        with self.store.session_scope():
+            # there must be states after the import
+            states = self.store.session.query(self.store.db_classes.State).all()
+            self.assertEqual(len(states), 1)
+
+            # there must be a single platform after the import
+            platforms = self.store.session.query(self.store.db_classes.Platform).all()
+            self.assertEqual(len(platforms), 1)
+
+            # there must be two datafiles afterwards - two files being processed
```
Comment (IanMayo) [inline tests/test_load_aircraft_csv.py:59]: comment looks wrong - we only processed one datafile

File: tests/test_load_aircraft_csv.py [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+            datafiles = self.store.session.query(self.store.db_classes.Datafile).all()
+            self.assertEqual(len(datafiles), 1)
+
+            # Check that there is an elevation of 147 reported (test file was manually edited
```
Comment (IanMayo) [inline tests/test_load_aircraft_csv.py:63]: file contains value of 1215 (ft).  The database stores data in SI units.  So, when we get the value from the database it should be the metres equivalent of 1215.

File: tests/test_load_aircraft_csv.py [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+            # to contain an elevation of 147m)
+            results = (
+                self.store.session.query(self.store.db_classes.State)
+                .filter(self.store.db_classes.State.elevation == 1215)
+                .all()
+            )
+            assert len(results) == 1
+            print(results[0].time)
+            assert results[0].time == datetime.datetime(2020, 3, 15, 10, 9, 0)
+
```
Comment (IanMayo) [inline tests/test_load_aircraft_csv.py:73]: Sam - there is only one state in the `States` table. So, you don't need to run a query to find entries with an elevation of 1215.  You can just retrieve the single `State` in the `States` table. Then you can check the values match those in the input datafile.  Certainly include `time` (which you're already doing), plus elevation, location, course and speed.

File: tests/sample_data/track_files/csv_data/aircraft/OwnPos_UC_MissingData.csv [hunk 4]
```diff
@@ -0,0 +1,2 @@
+Date(Uk),Time(Z),Lat(DD:MM.MM),Long(DDD:MM.MM),Lat(DegN),Long(DegE),Altitude(ft),GndSpeed(kts),Course(deg)
```
Comment (IanMayo) [inline tests/sample_data/track_files/csv_data/aircraft/OwnPos_UC_MissingData.csv:1]: @SamWilliams-dev - this file is called `missing-data`, but I can't see which field is missing.  It appears to be identical to `OwnPos_UC.csv`.

File: tests/test_load_aircraft_csv.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+from pepys_import.core.store.data_store import DataStore
+from pepys_import.file.file_processor import FileProcessor
+from tests.utils import check_errors_for_file_contents
+
+FILE_PATH = os.path.dirname(__file__)
+DATA_PATH_SHORT = os.path.join(FILE_PATH, "sample_data/track_files/csv_data/aircraft/OwnPos_UC.csv")
+DATA_PATH_LONG = os.path.join(
+    FILE_PATH, "sample_data/track_files/csv_data/aircraft/OwnPos_UC_Long.csv"
+)
+DATA_PATH_MISSING = os.path.join(
```
Comment (IanMayo) [inline tests/test_load_aircraft_csv.py:15]: DATA_PATH_MISSING isn't used. It appears you aren't testing the file for missing data.

File: importers/aircraft_csv_format_importer.py [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+            # Skip line if invalid timestamp
+            self.errors.append(
+                {self.error_type: f"Line {line_number}. Error in timestamp parsing."}
+            )
+            return
+
+        # And finally store it - Check how to do this as currently don't have a platform name
+        platform = self.get_cached_platform(data_store, platform_name=None, change_id=change_id)
+        sensor_type = data_store.add_to_sensor_types("Location-Satellite", change_id=change_id).name
+        privacy = get_lowest_privacy(data_store)
```
Comment (IanMayo) [inline importers/aircraft_csv_format_importer.py:95]: We can't use the lowest privacy here, we need to get it from the analyst (via the resolver)

File: importers/aircraft_csv_format_importer.py [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        lon_success = location.set_longitude_decimal_degrees(long_degrees_token.text)
+        if lat_success and lon_success:
+            state.location = location
+            combine_tokens(long_degrees_token, lat_degrees_token).record(
+                self.name, "location", state.location, "decimal degrees"
+            )
+        else:
+            self.errors.append({self.error_type: f"Line {line_number}. Error in Location Parsing."})
+
+        # Convert the altitude and check valid
```
Comment (IanMayo) [inline importers/aircraft_csv_format_importer.py:119]: the CSV header says the altitude is in feet

File: tests/test_load_aircraft_csv.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+        check_errors_for_file_contents(
+            "Date(Uk), \n, Blah, Blah",
+            "Not enough tokens:",
+            aircraft_importer,
+            filename="OwnPos_UC.csv",
+        )
+
+        # Test Invalid Date format
+        check_errors_for_file_contents(
+            "Date(Uk), \n, 20/03/15, 10:09:00, 50:37.12N, 005:06.24W, 50.61867, -5.104, 1215, 81.3, 311",
```
Comment (IanMayo) [inline tests/test_load_aircraft_csv.py:132]: Hmm, doesn't the `,` after `\n` mean the data line will start with a leading comma?

File: importers/aircraft_csv_format_importer.py (hunk 6 above)
Comment (SamWilliams-dev) [reply importers/aircraft_csv_format_importer.py:95]: Hi Ian, 
Currently all other importers either use this to set function above to set the privacy or they just don't appear to set it at all. The current exception that I can find is the `nmea_importer.py` which uses the following instead: `default_privacy="Private"`
Do you have any suggestions on how best to handle this?

File: tests/test_load_aircraft_csv.py (hunk 5 above)
Comment (SamWilliams-dev) [reply tests/test_load_aircraft_csv.py:15]: This was accidentally left over from my initial plan but was replaced by calls to `check_errors_for_file_contents` which now checks for an incorrect or missing format in each of the cell types to ensure that they're handled correctly. As a result, I don't think this is needed anymore and so I can probably remove it, unless you think it's still worth testing on a document as well as individual lines?

**Comment by IanMayo** [comment]
Morning @SamWilliams-dev - there are also two static analysis issues in the PR:
image

File: importers/aircraft_csv_format_importer.py (hunk 6 above)
Comment (robintw) [reply importers/aircraft_csv_format_importer.py:93]: You should be able to just pass `privacy=None` to the `get_sensor` call, and the resolver will ask the user to enter a privacy. (Please test this by running pepys_import _without_ `--resolver default` to check it asks for a privacy, as I may be wrong).

**Comment by robintw** [comment]
@SamWilliams-dev I've made a couple of responses to the comments that Ian made. Could you click the 'Mark as resolved' button for the ones you've fixed, so we can see which ones are still waiting to be done.

As for the static analysis comments about 'method could be a function' or 'method could be static' - I would suggest ignoring those. It is giving the message because there is a method on a class that doesn't actually use any of the other attributes/methods in that class, and therefore could be a standalone function. In this case, it makes sense to keep all of the test methods for the import inside the test class even if a couple of the methods don't _need_ to be in there - it just keeps everything organised nicely - so, I'd just ignore that. (You'll still need to deal with a docstring for the other warning though).

**Comment by IanMayo** [comment]
>  In this case, it makes sense to keep all of the test methods for the import inside the test class even if a couple of the methods don't _need_ to be in there - it just keeps everything organised nicely - so, I'd just ignore that. 

Is there an inline comment we can use to disable the warning just for that line?  That will show that we're consciously suppressing the warning for that line of code - and we haven't overlooked the waning.

**Comment by robintw** [comment]
Ah yes, it seems the warning comes from pylint, and can be suppressed by adding the following line as the first line of the method:

```
# pylint: disable=no-self-use
```

I'd add a comment above that line explaining why we're disabling it too.

**Comment by robintw** [comment]
Also @SamWilliams-dev, could you rename this PR so its name reflects the name of the importer, and that we're not doing a base class now

File: tests/test_load_aircraft_csv.py (hunk 5 above)
Comment (robintw) [reply tests/test_load_aircraft_csv.py:15]: I think that just using check_errors_for_file_contents is fine

File: importers/aircraft_csv_format_importer.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+        return True
+
+    def _load_this_line(self, data_store, line_number, line, datafile, change_id):
+        # Skip the header
+        if line_number == 1:
+            return
+
+        tokens = line.tokens(line.CSV_TOKENISER, ",")
+        if len(tokens) <= 1:
+            # The last line may be empty, don't worry
```
Comment (robintw) [inline importers/aircraft_csv_format_importer.py:42]: This if statement doesn't seem to be tested. Could you add a test that has a blank line at the end of the file, which should trigger this if?

File: importers/aircraft_csv_format_importer.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+            return
+        if len(tokens) < 8:
```
Comment (robintw) [inline importers/aircraft_csv_format_importer.py:44]: This tests for too few tokens, but what if there are too many? Should we be checking for exactly 9?

(Some of the other importers don't check for an exact number of tokens, as the lines in some formats can have a free-text comment at the end, which means the total number of tokens can be very high - but that doesn't apply here)

File: importers/aircraft_csv_format_importer.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+        altitude_token = tokens[6]
+        speed_token = tokens[7]
+        course_token = tokens[8]
+
+        # Check Date is in correct format
+        if len(date_token.text) != 10:
+            self.errors.append(
+                {
+                    self.error_type: f"Error on line {line_number}. Date format '{date_token.text}' "
+                    f"should be 10 figure data"
```
Comment (robintw) [inline importers/aircraft_csv_format_importer.py:65]: Do you mean `10 figure date`?

File: importers/aircraft_csv_format_importer.py [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+                }
+            )
+            return
+
+        timestamp = self.parse_timestamp(date_token.text, time_token.text)
+        if timestamp:
+            combine_tokens(date_token, time_token).record(self.name, "timestamp", timestamp)
+        else:
+            # Skip line if invalid timestamp
+            self.errors.append(
```
Comment (robintw) [inline importers/aircraft_csv_format_importer.py:85]: You seem to be missing a test for a timestamp that doesn't parse correctly - so this if isn't being covered by tests. I think you've got a test for the time being the wrong length (ie. not 8 char time), but not for an invalid date/time like 01:99:00 or 99/03/2015 etc)

File: importers/aircraft_csv_format_importer.py [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+                {self.error_type: f"Line {line_number}. Error in timestamp parsing."}
+            )
+            return
+
+        # And finally store it - Check how to do this as currently don't have a platform name
```
Comment (robintw) [inline importers/aircraft_csv_format_importer.py:90]: Comment needs removing/updating

File: importers/aircraft_csv_format_importer.py [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+        platform = self.get_cached_platform(data_store, platform_name=None, change_id=change_id)
```
Comment (robintw) [inline importers/aircraft_csv_format_importer.py:91]: Giving a platform_name of None here actually causes the importer to try and ask the user for the platform name for each row of the file - I've just tested it. I'll screenshare with you in a bit to talk through the solution.

File: importers/aircraft_csv_format_importer.py [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+            sensor_name="GPS",
+            sensor_type=sensor_type,
+            privacy=privacy,
+            change_id=change_id,
+        )
+        state = datafile.create_state(data_store, platform, sensor, timestamp, self.short_name)
+
+        # Set the location conversion
+        location = Location(errors=self.errors, error_type=self.error_type)
+        print(lat_degrees_token.text)
```
Comment (robintw) [inline importers/aircraft_csv_format_importer.py:105]: Why are we still printing the values here?

File: importers/aircraft_csv_format_importer.py [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+            course_token.record(self.name, "course", course)
+
+    @staticmethod
+    def parse_timestamp(date, time):
+        format_str = "%d/%m/%Y "
+        format_str += "%H:%M:%S"
+
+        try:
+            res = datetime.strptime(date.strip() + " " + time.strip(), format_str)
+        except ValueError:
```
Comment (robintw) [inline importers/aircraft_csv_format_importer.py:152]: This bit was missing coverage too, but should be caught by the test I suggested adding in one of my other comments - for something like a time of 12:99:30 or 99:01:02.

File: tests/sample_data/track_files/csv_data/aircraft/OwnPos_UC_MissingData.csv (hunk 4 above)
Comment (robintw) [reply tests/sample_data/track_files/csv_data/aircraft/OwnPos_UC_MissingData.csv:1]: Is this file being deleted now, as we're dealing with the missing/invalid data tests in a different way?

File: tests/test_load_aircraft_csv.py [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+
+            # there must be a single platform after the import
+            platforms = self.store.session.query(self.store.db_classes.Platform).all()
+            self.assertEqual(len(platforms), 1)
+
+            # there should only be one data file at the end of the test
+            datafiles = self.store.session.query(self.store.db_classes.Datafile).all()
+            self.assertEqual(len(datafiles), 1)
+
+            # Check that there is an elevation of 1215 reported (test file was manually edited
```
Comment (robintw) [inline tests/test_load_aircraft_csv.py:63]: Was it manually edited to contain an elevation of 1215ft? Or was it there originally? I think this is just a copy-paste

File: tests/test_load_aircraft_csv.py [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+            # to contain an elevation of 1215 ft)
+            results = (
+                self.store.session.query(self.store.db_classes.State)
+                # .filter(self.store.db_classes.State.elevation == 1215)
```
Comment (robintw) [inline tests/test_load_aircraft_csv.py:67]: Remove commented out code

File: tests/test_load_aircraft_csv.py [hunk 19]
```diff
@@ -0,0 +1,183 @@
...
+                .all()
+            )
+            assert len(results) == 1
+            assert results[0].time == datetime.datetime(2020, 3, 15, 10, 9, 0)
+            assert str(results[0].elevation) == "370.33199999999994 meter"
```
Comment (robintw) [inline tests/test_load_aircraft_csv.py:72]: The best way to test floating point values in a test is _not_ by converting them to a string! You can use the `round` function to round to a specific number of decimal places, and can check units by comparing to a value that has units.

For example:

```
round(value, 2) == 5.23 * unit_registry.metre
```

File: tests/test_load_aircraft_csv.py [hunk 20]
```diff
@@ -0,0 +1,183 @@
...
+            assert str(results[0].course) == "311.0 degree"
+            assert str(results[0].speed) == "41.824333333333335 meter / second"
+            assert str(results[0].location) == "-5.104, 50.619"
```
Comment (robintw) [inline tests/test_load_aircraft_csv.py:75]: Again, please don't use string comparison. See `test_load_nisida.py` L74 for an example of checking a longitude value

**Review by robintw — CHANGES_REQUESTED** [review]
Quite a few comments here Sam. Most are fairly simple, but I'll speak to you in a moment about the platform caching issue.

File: tests/sample_data/track_files/csv_data/aircraft/OwnPos_UC_MissingData.csv (hunk 4 above)
Comment (SamWilliams-dev) [reply tests/sample_data/track_files/csv_data/aircraft/OwnPos_UC_MissingData.csv:1]: Deleted this file now.

**Comment by IanMayo** [comment]
@SamWilliams-dev - I suggest the PR name is `Aircraft CSV Format Importer`.

We haven't followed a `base functionality` strategy. We've created the full implementation.

**Review by robintw — CHANGES_REQUESTED** [review]
Looks good now, thanks Sam.

One thing though: could you update the text of the PR (the text in the first 'comment' item on the PR page) to be up-to-date and correct. It currently says it's a WIP and that tests don't pass, and there isn't a list of what's been done etc. PRs are our long-term record of what has been done to the codebase and why, and we can look back at them to see why and how a change was made - so it's important this text is up-to-date.

**Comment by IanMayo** [comment]
> Looks good now, thanks Sam.
> 
> One thing though: could you update the text of the PR (the text in the first 'comment' item on the PR page) to be up-to-date and correct. It currently says it's a WIP and that tests don't pass, and there isn't a list of what's been done etc. PRs are our long-term record of what has been done to the codebase and why, and we can look back at them to see why and how a change was made - so it's important this text is up-to-date.

Thanks @robintw - you've made a good observation there.  We have to push to maintain the culture of the team/project - and be aware that we're developing _in the open_, and demonstrate our approach to quality.

**Review by IanMayo — APPROVED** [review]
Codacy error is a known issue with the the `pep257` linter:
https://stackoverflow.com/questions/45990301/pep257-d212-and-d213-conflicts

So, @robintw has advised to merge, and solve the test conflict in a separate PR.

**Pull request merged**